src/backend_model.cc

```diff
@@ -1367,36 +1367,40 @@ TRITONBACKEND_StateBuffer(
   SequenceState* to = reinterpret_cast<SequenceState*>(state);
   Status status = Status::Success;
 
-  // If the buffer size exactly matches the buffer available, reuse the
-  // currently allocated buffer.
-  if (to->Data()->TotalByteSize() == buffer_byte_size) {
-    const std::shared_ptr<AllocatedMemory>& memory =
-        reinterpret_cast<const std::shared_ptr<AllocatedMemory>&>(to->Data());
-
-    TRITONSERVER_MemoryType current_memory_type;
-    int64_t current_memory_type_id;
-    void* lbuffer =
-        memory->MutableBuffer(&current_memory_type, &current_memory_type_id);
-
-    // If the requested memory type doesn't match the current buffer, allocate
-    // a new buffer with the requested memory type and memory type id.
-    if (current_memory_type == *memory_type &&
-        current_memory_type_id == *memory_type_id) {
-      *buffer = lbuffer;
+  TRITONSERVER_MemoryType current_memory_type;
+  int64_t current_memory_type_id;
+  void* lbuffer = nullptr;
+  MutableMemory* mutable_memory =
+      reinterpret_cast<MutableMemory*>(to->Data().get());
+  lbuffer = mutable_memory->MutableBuffer(
+      &current_memory_type, &current_memory_type_id);
+
+  // If the buffer size exactly matches the buffer available and is requesting
+  // the same memory type and memory type id, reuse the currently allocated
+  // buffer.
+  if (to->Data()->TotalByteSize() == buffer_byte_size &&
+      current_memory_type == *memory_type &&
+      current_memory_type_id == *memory_type_id) {
+    *buffer = lbuffer;
+  } else {
+    if (to->UseGrowableMemory()) {
+      GrowableMemory* growable_memory =
+          reinterpret_cast<GrowableMemory*>(to->Data().get());
+      RETURN_TRITONSERVER_ERROR_IF_ERROR(
+          growable_memory->Resize(buffer_byte_size));
+      *buffer = growable_memory->MutableBuffer(memory_type, memory_type_id);
     } else {
       std::shared_ptr<AllocatedMemory> memory =
           std::make_shared<AllocatedMemory>(
               buffer_byte_size, *memory_type, *memory_type_id);
       *buffer = memory->MutableBuffer(memory_type, memory_type_id);
       to->RemoveAllData();
       status = to->SetData(memory);
+      if (to->ReuseBuffer()) {
+        to->OtherState()->RemoveAllData();
+        to->OtherState()->SetData(memory);
+      }
```

Comment:
Why isn't this section needed in UseGrowableMemory condition too? Doesn't growable memory re-use buffer by definition?

Reply:
Growable memory uses the same "memory" for both input and output state. Since in Growable memory we're not creating a new buffer it is already pointing to the same buffer so this is not required.

Reply:
Two questions:
1) Can we expand growable memory to cases where input and output are different buffers? Seems like the two are independent -
2) in the case we don't use the same buffer - do we have to also increase the memory for both? (Is that handled somewhere else?)

Reply:
I think growable memory doesn't provide a lot of value if you are not using a single buffer. The main goal of the growable memory is to provide speedups for the cases where you're able to append state to the existing state object. When you're using using a single state buffer and the state size changes, you have to copy the whole state even if you have an option to grow an existing allocation. I think in that scenario, just getting a new allocation isn't any different than getting a new buffer from user's perspective.

Reply:
won't it cut down memory copies in a similar way to single buffer?

1) Request larger state buffer size
2) append data
3) switch input and output buffers
4) append data to output buffer for next update